src/ui/components/RecipeEnrollmentPage.tsx

```diff
@@ -1,38 +1,129 @@
+import { NimbusExperiment } from "@mozilla/nimbus-schemas";
 
 import { useToastsContext } from "../hooks/useToasts";
 
 const RecipeEnrollmentPage: FC = () => {
   const [jsonInput, setJsonInput] = useState("");
+  const [enrollError, setEnrollError] =
+    useState<EnrollInExperimentResult["error"]>(null);
   const { addToast } = useToastsContext();
 
+  const recipe = useMemo<NimbusExperiment | null>(() => {
+    try {
+      return JSON.parse(jsonInput) as NimbusExperiment;
+    } catch {
+      return null;
+    }
+  }, [jsonInput]);
+
+  const slug = recipe?.slug || "";
+
   const handleInputChange = useCallback(
     (event: ChangeEvent<HTMLTextAreaElement>) => {
       setJsonInput(event.target.value);
     },
-    [],
+    [setJsonInput],
   );
 
-  const handleEnrollClick = useCallback(async () => {
-    try {
-      const result = await browser.experiments.nimbus.enrollInExperiment(
-        JSON.parse(jsonInput) as object,
-      );
+  const handleEnrollClick = useCallback(
+    async (
+      event?: React.MouseEvent<HTMLButtonElement>,
+      forceEnroll = false,
+    ) => {
+      if (!jsonInput) {
+        addToast({ message: "Invalid Input: Enter JSON", variant: "danger" });
+        return;
+      }
 
-      if (result) {
-        addToast({ message: "Enrollment successful", variant: "success" });
-        setJsonInput("");
-      } else {
-        addToast({ message: "Enrollment failed", variant: "danger" });
+      let parsedRecipe;
+      try {
+        parsedRecipe = JSON.parse(jsonInput) as object;
+      } catch (error) {
+        addToast({
+          message: `Error enrolling into experiment: ${
+            (error as Error).message ?? String(error)
+          }`,
+          variant: "danger",
+        });
+        return;
       }
-    } catch (error) {
-      addToast({
-        message: `Error enrolling into experiment: ${(error as Error).message ?? String(error)}`,
-        variant: "danger",
-      });
+
+      try {
+        const result = await browser.experiments.nimbus.enrollInExperiment(
+          parsedRecipe as object,
+          forceEnroll,
+        );
+
+        if (result.enrolled) {
+          addToast({ message: "Enrollment successful", variant: "success" });
+          setJsonInput("");
+        } else if (result.error) {
+          setEnrollError(result.error);
+        }
+      } catch (error) {
+        addToast({
+          message: `Error enrolling into experiment: ${
+            (error as Error).message ?? String(error)
+          }`,
+          variant: "danger",
+        });
+      }
+    },
+    [jsonInput, setJsonInput, addToast, setEnrollError],
+  );
+
+  const handleModalConfirm = useCallback(async () => {
+    setEnrollError(null);
+    await handleEnrollClick(undefined, true);
+  }, [handleEnrollClick, setEnrollError]);
+
+  const handleModalClose = useCallback(() => {
+    setEnrollError(null);
+  }, [setEnrollError]);
+
+  function EnrollmentError(
```

Comment:
This should be a top level function, not nested inside the component fn

if its nested inside, it will always trigger a re-render because the function is redefined every time the component is called